File: iosApp/iosApp/Presentation/Main/CafeManagement/MenuGoods/MenuGoodsViewModel.swift
//
//  MenuGoodsViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/11.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class MenuGoodsViewModel: ObservableObject {
    private let cafeId: String

    private let getCafeDetailUseCase: GetCafeDetailUseCase

    private let upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase

    private let deleteCafeMenuGoodsUseCase: DeleteCafeMenuGoodsUseCase

    private let cafeDetailEventPublisher: CafeDetailEventPublisher

    @Published private(set) var uiState = MenuGoodsUiState()

    let event = PassthroughSubject<MenuGoodsEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func loadMenuGoods() {
        tasks[.load]?.cancel()
        uiState.isLoading = true
        uiState.infoMessage = nil

        tasks[.load] = Task { [weak self] in
            guard let self else { return }

            do {
                let result = try await getCafeDetailUseCase.invoke(cafeId: cafeId)
                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? CafeDetailFeed {
                    applyDetail(feed.detail)
                } else {
                    uiState.isLoading = false
                    uiState.infoMessage = "항목 정보를 불러오지 못했습니다."
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.infoMessage = "항목 정보를 불러오지 못했습니다."
            }
        }
    }

    private func observeCafeDetailEvent() {
        tasks[.detailEvent]?.cancel()
        tasks[.detailEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeDetailEventPublisher.events) {
                    switch event {
                    case let event as CafeDetailEvent.CafeInfoUpdated:
                        if event.cafeId == self.cafeId {
                            self.loadMenuGoods()
                        }
                    case let event as CafeDetailEvent.MenuCreated:
                        if event.cafeId == self.cafeId {
                            self.loadMenuGoods()
                        }
                    case let event as CafeDetailEvent.MenuUpdated:
                        if event.cafeId == self.cafeId {
                            self.upsertLocalMenu(event.menu)
                        }
                    case let event as CafeDetailEvent.MenuDeleted:
                        if event.cafeId == self.cafeId {
                            self.removeLocalMenu(itemId: event.itemId)
                        }
                    case let event as CafeDetailEvent.GoodsCreated:
                        if event.cafeId == self.cafeId {
                            self.loadMenuGoods()
                        }
                    case let event as CafeDetailEvent.GoodsUpdated:
                        if event.cafeId == self.cafeId {
                            self.upsertLocalGoods(event.goods)
                        }
                    case let event as CafeDetailEvent.GoodsDeleted:
                        if event.cafeId == self.cafeId {
                            self.removeLocalGoods(itemId: event.itemId)
                        }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func applyDetail(_ detail: CafeDetail) {
        var nextState = self.uiState
        nextState.cafeName = detail.cafe.name
        nextState.isLoading = false
        nextState.menuCategories = self.buildMenuCategories(items: detail.menus)
        nextState.goodsCategories = self.buildGoodsCategories(items: detail.goods)
        nextState.menuItems = detail.menus
        nextState.goodsItems = detail.goods
        nextState.menuAvailabilityOverrides = [:]
        nextState.goodsAvailabilityOverrides = [:]
        nextState.infoMessage = nil
        self.uiState = nextState
    }

    private func upsertLocalMenu(_ menu: CafeMenu) {
        var nextState = uiState
        let nextMenuItems = nextState.menuItems.filter { $0.id != menu.id } + [menu]
        let nextGoodsItems = nextState.goodsItems.filter { $0.id != menu.id }

        nextState.isLoading = false
        nextState.menuCategories = buildMenuCategories(items: nextMenuItems)
        nextState.goodsCategories = buildGoodsCategories(items: nextGoodsItems)
        nextState.menuItems = nextMenuItems
        nextState.goodsItems = nextGoodsItems
        nextState.menuAvailabilityOverrides.removeValue(forKey: menu.id)
        nextState.goodsAvailabilityOverrides.removeValue(forKey: menu.id)
        if nextState.pendingDeleteItemId == menu.id { nextState.pendingDeleteItemId = nil }
        uiState = nextState
    }

    private func upsertLocalGoods(_ goods: Goods) {
        var nextState = uiState
        let nextMenuItems = nextState.menuItems.filter { $0.id != goods.id }
        let nextGoodsItems = nextState.goodsItems.filter { $0.id != goods.id } + [goods]
        nextState.isLoading = false
        nextState.menuCategories = buildMenuCategories(items: nextMenuItems)
        nextState.goodsCategories = buildGoodsCategories(items: nextGoodsItems)
        nextState.menuItems = nextMenuItems
        nextState.goodsItems = nextGoodsItems
        nextState.menuAvailabilityOverrides.removeValue(forKey: goods.id)
        nextState.goodsAvailabilityOverrides.removeValue(forKey: goods.id)
        if nextState.pendingDeleteItemId == goods.id { nextState.pendingDeleteItemId = nil }
        uiState = nextState
    }

    private func removeLocalMenu(itemId: String) {
        var nextState = uiState
        let nextMenuItems = nextState.menuItems.filter { $0.id != itemId }
        nextState.isLoading = false
        nextState.menuCategories = buildMenuCategories(items: nextMenuItems)
        nextState.goodsCategories = buildGoodsCategories(items: nextState.goodsItems)
        nextState.menuItems = nextMenuItems
        nextState.menuAvailabilityOverrides.removeValue(forKey: itemId)
        nextState.goodsAvailabilityOverrides.removeValue(forKey: itemId)
        if nextState.pendingDeleteItemId == itemId { nextState.pendingDeleteItemId = nil }
        uiState = nextState
    }

    private func removeLocalGoods(itemId: String) {
        var nextState = uiState
        let nextGoodsItems = nextState.goodsItems.filter { $0.id != itemId }
        nextState.isLoading = false
        nextState.menuCategories = buildMenuCategories(items: nextState.menuItems)
        nextState.goodsCategories = buildGoodsCategories(items: nextGoodsItems)
        nextState.goodsItems = nextGoodsItems
        nextState.menuAvailabilityOverrides.removeValue(forKey: itemId)
        nextState.goodsAvailabilityOverrides.removeValue(forKey: itemId)
        if nextState.pendingDeleteItemId == itemId { nextState.pendingDeleteItemId = nil }
        uiState = nextState
    }

    private func toggleSearch() {
        uiState.isSearchVisible.toggle()
        if !uiState.isSearchVisible {
            uiState.searchQuery = ""
        }
    }

    private func selectCategory(_ categoryId: String?) {
        switch uiState.selectedCollection {
        case .menu:
            uiState.selectedMenuCategoryId = categoryId
        case .goods:
            uiState.selectedGoodsCategoryId = categoryId
        }
    }

    private func toggleItemAvailability(_ itemId: String) {
        switch uiState.selectedCollection {
        case .menu:
            guard let item = uiState.menuItems.first(where: { $0.id == itemId }) else { return }
            let nextAvailability = !uiState.isMenuAvailable(item)
            uiState.menuAvailabilityOverrides[itemId] = nextAvailability
            uiState.infoMessage = nil
            tasks[.toggleAvailability]?.cancel()
            tasks[.toggleAvailability] = Task { [weak self] in
                guard let self else { return }
                do {
                    let result = try await upsertCafeMenuGoodsUseCase.invoke(
                        update: CafeMenuGoodsUpsert(
                            cafeId: cafeId,
                            itemId: item.id,
                            name: item.name,
                            price: item.price,
                            category: item.category,
                            description: item.desc,
                            isInStock: nextAvailability,
                            imageUrl: item.image
                        )
                    )
                    if result is AppResultFailure {
                        uiState.menuAvailabilityOverrides.removeValue(forKey: itemId)
                        uiState.infoMessage = "판매 상태 저장에 실패했습니다."
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.menuAvailabilityOverrides.removeValue(forKey: itemId)
                    uiState.infoMessage = "판매 상태 저장에 실패했습니다."
                }
            }
        case .goods:
            guard let item = uiState.goodsItems.first(where: { $0.id == itemId }) else { return }
            let nextAvailability = !uiState.isGoodsAvailable(item)
            uiState.goodsAvailabilityOverrides[itemId] = nextAvailability
            uiState.infoMessage = nil
            tasks[.toggleAvailability]?.cancel()
            tasks[.toggleAvailability] = Task { [weak self] in
                guard let self else { return }
                do {
                    let result = try await upsertCafeMenuGoodsUseCase.invoke(
                        update: CafeMenuGoodsUpsert(
                            cafeId: cafeId,
                            itemId: item.id,
                            name: item.name,
                            price: item.price,
                            category: "goods",
                            description: "카페 굿즈 판매 항목",
                            isInStock: nextAvailability,
                            imageUrl: item.image
                        )
                    )
                    if result is AppResultFailure {
                        uiState.goodsAvailabilityOverrides.removeValue(forKey: itemId)
                        uiState.infoMessage = "판매 상태 저장에 실패했습니다."
                    }
                } catch {
                    if Task.isCancelled { return }
                    uiState.goodsAvailabilityOverrides.removeValue(forKey: itemId)
                    uiState.infoMessage = "판매 상태 저장에 실패했습니다."
                }
            }
        }
    }

    private func deleteItem(_ itemId: String) {
        uiState.pendingDeleteItemId = itemId
    }

    private func confirmDeleteItem(_ itemId: String) {
        uiState.infoMessage = nil
        uiState.pendingDeleteItemId = nil
        Task { [weak self] in
            guard let self else { return }

            do {
                let result = try await deleteCafeMenuGoodsUseCase.invoke(cafeId: cafeId, itemId: itemId)
                if result is AppResultSuccess<AnyObject> {
                    uiState.infoMessage = "항목이 삭제되었습니다."
                } else {
                    uiState.infoMessage = "항목 삭제에 실패했습니다."
                }
            } catch {
                if Task.isCancelled { return }
                uiState.infoMessage = "항목 삭제에 실패했습니다."
            }
        }
    }

    private func cancelDeleteItem() {
        uiState.pendingDeleteItemId = nil
    }

    private func buildMenuCategories(items: [CafeMenu]) -> [MenuGoodsUiState.CategoryChip] {
        let preferredOrder: [(String, MenuGoodsUiState.CategoryChip)] = [
            ("food", .init(id: "food", label: "Food", iconKey: "food")),
            ("drink", .init(id: "drink", label: "Drinks", iconKey: "drink")),
            ("dessert", .init(id: "dessert", label: "Dessert", iconKey: "dessert"))
        ]
        return [.init(id: nil, label: "All", iconKey: "all")] + preferredOrder.compactMap { id, chip in
            items.contains(where: { $0.category.lowercased() == id }) ? chip : nil
        }
    }

    private func buildGoodsCategories(items: [Goods]) -> [MenuGoodsUiState.CategoryChip] {
        var seen = Set<String>()
        let dynamic = items.compactMap { item -> MenuGoodsUiState.CategoryChip? in
            let categoryId: String
            if item.name.localizedCaseInsensitiveContains("포토") {
                categoryId = "collectible"
            } else if item.name.localizedCaseInsensitiveContains("의상") {
                categoryId = "apparel"
            } else {
                categoryId = "goods"
            }
            guard !seen.contains(categoryId) else {
                return nil
            }
            seen.insert(categoryId)
            return MenuGoodsUiState.CategoryChip(
                id: categoryId,
                label: categoryId == "collectible" ? "Collectible" : (categoryId == "apparel" ? "Apparel" : "Goods"),
                iconKey: "goods"
            )
        }

        return [.init(id: nil, label: "All", iconKey: "all")] + dynamic
    }

    func onAction(_ action: MenuGoodsAction) {
        switch action {
        case .clickBack:
            event.send(.navigateBack)
        case .clickSearch:
            toggleSearch()
        case .changeSearchQuery(let value):
            uiState.searchQuery = value
        case .selectCollection(let collection):
            uiState.selectedCollection = collection
        case .selectCategory(let categoryId):
            selectCategory(categoryId)
        case .toggleItemAvailability(let itemId):
            toggleItemAvailability(itemId)
        case .clickEditItem(let itemId):
            event.send(.navigateToEdit(cafeId: cafeId, itemId: itemId))
        case .clickDeleteItem(let itemId):
            deleteItem(itemId)
        case .confirmDeleteItem(let itemId):
            confirmDeleteItem(itemId)
        case .cancelDeleteItem:
            cancelDeleteItem()
        case .clickAddNewItem:
            event.send(.navigateToEdit(cafeId: cafeId, itemId: nil))
        case .dismissInfoMessage:
            uiState.infoMessage = nil
        }
    }

    init(
        cafeId: String,
        getCafeDetailUseCase: GetCafeDetailUseCase = KoinInitializerKt.resolveGetCafeDetailUseCase(),
        upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase = KoinInitializerKt.resolveUpsertCafeMenuGoodsUseCase(),
        deleteCafeMenuGoodsUseCase: DeleteCafeMenuGoodsUseCase = KoinInitializerKt.resolveDeleteCafeMenuGoodsUseCase(),
        cafeDetailEventPublisher: CafeDetailEventPublisher = KoinInitializerKt.resolveCafeDetailEventPublisher()
    ) {
        self.cafeId = cafeId
        self.getCafeDetailUseCase = getCafeDetailUseCase
        self.upsertCafeMenuGoodsUseCase = upsertCafeMenuGoodsUseCase
        self.deleteCafeMenuGoodsUseCase = deleteCafeMenuGoodsUseCase
        self.cafeDetailEventPublisher = cafeDetailEventPublisher

        observeCafeDetailEvent()
        loadMenuGoods()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case load
        case detailEvent
        case toggleAvailability
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/menugoods/MenuGoodsViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.menugoods

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.CafeDetailEvent
import com.hhp227.concafe.domain.event.publisher.CafeDetailEventPublisher
import com.hhp227.concafe.domain.model.CafeDetail
import com.hhp227.concafe.domain.model.CafeMenu
import com.hhp227.concafe.domain.model.CafeMenuGoodsUpsert
import com.hhp227.concafe.domain.model.Goods
import com.hhp227.concafe.domain.usecase.DeleteCafeMenuGoodsUseCase
import com.hhp227.concafe.domain.usecase.GetCafeDetailUseCase
import com.hhp227.concafe.domain.usecase.UpsertCafeMenuGoodsUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class MenuGoodsViewModel(
    private val cafeId: String,
    private val getCafeDetailUseCase: GetCafeDetailUseCase,
    private val upsertCafeMenuGoodsUseCase: UpsertCafeMenuGoodsUseCase,
    private val deleteCafeMenuGoodsUseCase: DeleteCafeMenuGoodsUseCase,
    private val cafeDetailEventPublisher: CafeDetailEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(MenuGoodsUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<MenuGoodsEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<JobKey, Job>()

    private fun loadMenuGoods() {
        jobs[JobKey.LOAD]?.cancel()
        jobs[JobKey.LOAD] = viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, infoMessage = null) }
            when (val result = getCafeDetailUseCase.invoke(cafeId)) {
                is AppResult.Success -> applyDetail(result.data.detail)
                is AppResult.Failure -> {
                    _uiState.update { it.copy(isLoading = false, infoMessage = "항목 정보를 불러오지 못했습니다.") }
                }
            }
        }
    }

    private fun observeCafeDetailEvent() {
        jobs[JobKey.OBSERVE_EVENT]?.cancel()
        jobs[JobKey.OBSERVE_EVENT] = viewModelScope.launch {
            cafeDetailEventPublisher.events.collect { event ->
                when (event) {
                    is CafeDetailEvent.CafeInfoUpdated -> if (event.cafeId == cafeId) {
                        loadMenuGoods()
                    }
                    is CafeDetailEvent.MenuCreated -> if (event.cafeId == cafeId) {
                        loadMenuGoods()
                    }
                    is CafeDetailEvent.MenuUpdated -> if (event.cafeId == cafeId) {
                        upsertLocalMenu(event.menu)
                    }
                    is CafeDetailEvent.MenuDeleted -> if (event.cafeId == cafeId) {
                        removeLocalMenu(event.itemId)
                    }
                    is CafeDetailEvent.GoodsCreated -> if (event.cafeId == cafeId) {
                        loadMenuGoods()
                    }
                    is CafeDetailEvent.GoodsUpdated -> if (event.cafeId == cafeId) {
                        upsertLocalGoods(event.goods)
                    }
                    is CafeDetailEvent.GoodsDeleted -> if (event.cafeId == cafeId) {
                        removeLocalGoods(event.itemId)
                    }
                }
            }
        }
    }

    private fun applyDetail(detail: CafeDetail) {
        _uiState.update {
            it.copy(
                cafeName = detail.cafe.name,
                isLoading = false,
                menuCategories = buildMenuCategories(detail.menus),
                goodsCategories = buildGoodsCategories(detail.goods),
                menuItems = detail.menus,
                goodsItems = detail.goods,
                menuAvailabilityOverrides = emptyMap(),
                goodsAvailabilityOverrides = emptyMap(),
                infoMessage = null
            )
        }
    }

    private fun upsertLocalMenu(menu: CafeMenu) {
        _uiState.update { state ->
            val nextMenuItems = state.menuItems.filterNot { it.id == menu.id } + menu
            val nextGoodsItems = state.goodsItems.filterNot { it.id == menu.id }
            state.copy(
                isLoading = false,
                menuCategories = buildMenuCategories(nextMenuItems),
                goodsCategories = buildGoodsCategories(nextGoodsItems),
                menuItems = nextMenuItems,
                goodsItems = nextGoodsItems,
                menuAvailabilityOverrides = state.menuAvailabilityOverrides - menu.id,
                goodsAvailabilityOverrides = state.goodsAvailabilityOverrides - menu.id,
                pendingDeleteItemId = state.pendingDeleteItemId.takeUnless { it == menu.id }
            )
        }
    }

    private fun upsertLocalGoods(goods: Goods) {
        _uiState.update { state ->
            val nextMenuItems = state.menuItems.filterNot { it.id == goods.id }
            val nextGoodsItems = state.goodsItems.filterNot { it.id == goods.id } + goods
            state.copy(
                isLoading = false,
                menuCategories = buildMenuCategories(nextMenuItems),
                goodsCategories = buildGoodsCategories(nextGoodsItems),
                menuItems = nextMenuItems,
                goodsItems = nextGoodsItems,
                menuAvailabilityOverrides = state.menuAvailabilityOverrides - goods.id,
                goodsAvailabilityOverrides = state.goodsAvailabilityOverrides - goods.id,
                pendingDeleteItemId = state.pendingDeleteItemId.takeUnless { it == goods.id }
            )
        }
    }

    private fun removeLocalMenu(itemId: String) {
        _uiState.update { state ->
            val nextMenuItems = state.menuItems.filterNot { it.id == itemId }
            state.copy(
                isLoading = false,
                menuCategories = buildMenuCategories(nextMenuItems),
                goodsCategories = buildGoodsCategories(state.goodsItems),
                menuItems = nextMenuItems,
                goodsItems = state.goodsItems,
                menuAvailabilityOverrides = state.menuAvailabilityOverrides - itemId,
                goodsAvailabilityOverrides = state.goodsAvailabilityOverrides - itemId,
                pendingDeleteItemId = state.pendingDeleteItemId.takeUnless { it == itemId }
            )
        }
    }

    private fun removeLocalGoods(itemId: String) {
        _uiState.update { state ->
            val nextGoodsItems = state.goodsItems.filterNot { it.id == itemId }
            state.copy(
                isLoading = false,
                menuCategories = buildMenuCategories(state.menuItems),
                goodsCategories = buildGoodsCategories(nextGoodsItems),
                menuItems = state.menuItems,
                goodsItems = nextGoodsItems,
                menuAvailabilityOverrides = state.menuAvailabilityOverrides - itemId,
                goodsAvailabilityOverrides = state.goodsAvailabilityOverrides - itemId,
                pendingDeleteItemId = state.pendingDeleteItemId.takeUnless { it == itemId }
            )
        }
    }

    private fun clickBack() {
        viewModelScope.launch {
            _event.emit(MenuGoodsEvent.NavigateBack)
        }
    }

    private fun toggleSearch() {
        _uiState.update { state ->
            state.copy(
                isSearchVisible = !state.isSearchVisible,
                searchQuery = if (state.isSearchVisible) "" else state.searchQuery
            )
        }
    }

    private fun changeSearchQuery(value: String) {
        _uiState.update { it.copy(searchQuery = value) }
    }

    private fun selectCollection(collection: MenuGoodsUiState.CollectionTab) {
        _uiState.update { it.copy(selectedCollection = collection) }
    }

    private fun selectCategory(categoryId: String?) {
        _uiState.update { state ->
            when (state.selectedCollection) {
                MenuGoodsUiState.CollectionTab.MENU -> state.copy(selectedMenuCategoryId = categoryId)
                MenuGoodsUiState.CollectionTab.GOODS -> state.copy(selectedGoodsCategoryId = categoryId)
            }
        }
    }

    private fun toggleItemAvailability(itemId: String) {
        when (_uiState.value.selectedCollection) {
            MenuGoodsUiState.CollectionTab.MENU -> {
                val target = _uiState.value.menuItems.firstOrNull { it.id == itemId } ?: return
                val nextAvailability = !_uiState.value.isMenuAvailable(target)
                _uiState.update { state ->
                    state.copy(
                        menuAvailabilityOverrides = state.menuAvailabilityOverrides + (itemId to nextAvailability),
                        infoMessage = null
                    )
                }
                jobs[JobKey.TOGGLE_AVAILABILITY]?.cancel()
                jobs[JobKey.TOGGLE_AVAILABILITY] = viewModelScope.launch {
                    when (
                        upsertCafeMenuGoodsUseCase.invoke(
                            CafeMenuGoodsUpsert(
                                cafeId = cafeId,
                                itemId = target.id,
                                name = target.name,
                                price = target.price,
                                category = target.category,
                                description = target.desc,
                                isInStock = nextAvailability,
                                imageUrl = target.image
                            )
                        )
                    ) {
                        is AppResult.Success -> Unit
                        is AppResult.Failure -> {
                            _uiState.update { state ->
                                state.copy(
                                    menuAvailabilityOverrides = state.menuAvailabilityOverrides - itemId,
                                    infoMessage = "판매 상태 저장에 실패했습니다."
                                )
                            }
                        }
                    }
                }
            }
            MenuGoodsUiState.CollectionTab.GOODS -> {
                val target = _uiState.value.goodsItems.firstOrNull { it.id == itemId } ?: return
                val nextAvailability = !_uiState.value.isGoodsAvailable(target)
                _uiState.update { state ->
                    state.copy(
                        goodsAvailabilityOverrides = state.goodsAvailabilityOverrides + (itemId to nextAvailability),
                        infoMessage = null
                    )
                }
                jobs[JobKey.TOGGLE_AVAILABILITY]?.cancel()
                jobs[JobKey.TOGGLE_AVAILABILITY] = viewModelScope.launch {
                    when (
                        upsertCafeMenuGoodsUseCase.invoke(
                            CafeMenuGoodsUpsert(
                                cafeId = cafeId,
                                itemId = target.id,
                                name = target.name,
                                price = target.price,
                                category = "goods",
                                description = "카페 굿즈 판매 항목",
                                isInStock = nextAvailability,
                                imageUrl = target.image
                            )
                        )
                    ) {
                        is AppResult.Success -> Unit
                        is AppResult.Failure -> {
                            _uiState.update { state ->
                                state.copy(
                                    goodsAvailabilityOverrides = state.goodsAvailabilityOverrides - itemId,
                                    infoMessage = "판매 상태 저장에 실패했습니다."
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    private fun clickDeleteItem(itemId: String) {
        _uiState.update { it.copy(pendingDeleteItemId = itemId) }
    }

    private fun confirmDeleteItem(itemId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(infoMessage = null, pendingDeleteItemId = null) }
            when (deleteCafeMenuGoodsUseCase.invoke(cafeId = cafeId, itemId = itemId)) {
                is AppResult.Success -> {
                    _uiState.update { it.copy(infoMessage = "항목이 삭제되었습니다.") }
                }
                is AppResult.Failure -> {
                    _uiState.update { it.copy(infoMessage = "항목 삭제에 실패했습니다.") }
                }
            }
        }
    }

    private fun cancelDeleteItem() {
        _uiState.update { it.copy(pendingDeleteItemId = null) }
    }

    private fun dismissInfoMessage() {
        _uiState.update { it.copy(infoMessage = null) }
    }

    private fun clickAddNewItem() {
        viewModelScope.launch {
            _event.emit(MenuGoodsEvent.NavigateToEdit(cafeId = cafeId))
        }
    }

    private fun clickEditItem(itemId: String) {
        viewModelScope.launch {
            _event.emit(MenuGoodsEvent.NavigateToEdit(cafeId = cafeId, itemId = itemId))
        }
    }

    private fun buildMenuCategories(
        items: List<CafeMenu>
    ): List<MenuGoodsUiState.CategoryChip> {
        val preferredOrder = listOf(
            "food" to MenuGoodsUiState.CategoryChip("food", "Food", "food"),
            "drink" to MenuGoodsUiState.CategoryChip("drink", "Drinks", "drink"),
            "dessert" to MenuGoodsUiState.CategoryChip("dessert", "Dessert", "dessert")
        )
        return listOf(MenuGoodsUiState.CategoryChip(null, "All", "all")) +
            preferredOrder.mapNotNull { (id, chip) ->
                chip.takeIf { items.any { menu -> menu.category.lowercase() == id } }
            }
    }

    private fun buildGoodsCategories(
        items: List<Goods>
    ): List<MenuGoodsUiState.CategoryChip> {
        val dynamic = items
            .map { item ->
                val categoryId = when {
                    item.name.contains("포토", ignoreCase = true) -> "collectible"
                    item.name.contains("의상", ignoreCase = true) -> "apparel"
                    else -> "goods"
                }
                MenuGoodsUiState.CategoryChip(
                    id = categoryId,
                    label = when (categoryId) {
                        "collectible" -> "Collectible"
                        "apparel" -> "Apparel"
                        else -> "Goods"
                    },
                    iconKey = "goods"
                )
            }
            .distinctBy { it.id }
        return listOf(MenuGoodsUiState.CategoryChip(null, "All", "all")) + dynamic
    }

    fun onAction(action: MenuGoodsAction) {
        when (action) {
            MenuGoodsAction.ClickBack -> clickBack()
            MenuGoodsAction.ClickSearch -> toggleSearch()
            is MenuGoodsAction.ChangeSearchQuery -> changeSearchQuery(action.value)
            is MenuGoodsAction.SelectCollection -> selectCollection(action.collection)
            is MenuGoodsAction.SelectCategory -> selectCategory(action.categoryId)
            is MenuGoodsAction.ToggleItemAvailability -> toggleItemAvailability(action.itemId)
            is MenuGoodsAction.ClickEditItem -> clickEditItem(action.itemId)
            is MenuGoodsAction.ClickDeleteItem -> clickDeleteItem(action.itemId)
            is MenuGoodsAction.ConfirmDeleteItem -> confirmDeleteItem(action.itemId)
            MenuGoodsAction.CancelDeleteItem -> cancelDeleteItem()
            MenuGoodsAction.ClickAddNewItem -> clickAddNewItem()
            MenuGoodsAction.DismissInfoMessage -> dismissInfoMessage()
        }
    }

    init {
        observeCafeDetailEvent()
        loadMenuGoods()
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    private enum class JobKey {
        LOAD,
        OBSERVE_EVENT,
        TOGGLE_AVAILABILITY
    }
}
